[libpas] Fix the libpas tests build

dhecht opened the pull request:
#### 8f3d6b367eb5893dd4a9f2cab616732fc1e80bdf
<pre>
[libpas] Fix the libpas tests build
<a href="https://bugs.webkit.org/show_bug.cgi?id=293960">https://bugs.webkit.org/show_bug.cgi?id=293960</a>
<a href="https://rdar.apple.com/152507524">rdar://152507524</a>

Reviewed by Yijia Huang.

* Source/bmalloc/libpas/libpas.xcodeproj/project.pbxproj:

pas_thread.{c,h} was added in 295497@main so add it to the
xcode project so that the tests can be built on darwin.

* Source/bmalloc/libpas/src/libpas/pas_segregated_shared_view.h:
(pas_segregated_shared_view_compute_minimum_size_for_bump_with_aligned_padding):

This function was added in 293476@main. The compiler complains about the
implicit conversion when building the tests but not when building the product.
Add the cast to silence the warning  (promoted to error), like we do in other places.

OpenSource/Source/bmalloc/libpas/src/libpas/pas_segregated_shared_view.h:137:9: error:
      implicit conversion loses integer precision: &apos;uintptr_t&apos; (aka &apos;unsigned long&apos;) to &apos;unsigned int&apos;
      [-Werror,-Wshorten-64-to-32]
  136 |     return pas_round_up_to_power_of_2(allocation_size, trailing_padding_alignment)
      |     ~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  137 |         + trailing_padding_size;
      |         ^~~~~~~~~~~~~~~~~~~~~~~

Canonical link: <a href="https://commits.webkit.org/295762@main">https://commits.webkit.org/295762@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1a388703a256aa0da7075d54d93e1c66a527b4ed

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106060 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25809 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16203 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111257 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56656 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26465 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34312 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/80573 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109064 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20904 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95718 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60895 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/20487 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13816 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56094 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/98699 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90273 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13851 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114112 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/104677 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33198 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/24501 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/89654 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33562 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91946 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89343 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22769 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34202 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12014 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28769 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33123 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/128989 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32869 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35168 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36219 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34467 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/bmalloc/libpas/src/libpas/pas_segregated_shared_view.h:
```diff
@@ -133,7 +133,7 @@ pas_segregated_shared_view_compute_minimum_size_for_bump_with_aligned_padding(
     unsigned trailing_padding_alignment)
 {
-    return pas_round_up_to_power_of_2(allocation_size, trailing_padding_alignment)
+    return (unsigned)pas_round_up_to_power_of_2(allocation_size, trailing_padding_alignment)
```
Should we assert this?
```
uintptr_t size = pas_round_up_to_power_of_2(...);
PAS_ASSERT(size <= UINT_MAX);
return (unsigned)size + trailing_padding_size;
```

dhecht replied:
inputs are `unsigned` type so seems unnecessary but I can add if you still prefer.

hyjorc1 replied:
Yeah, one of the existing uses states that
```
    /* NOTE: This math cannot overflow because size and alignment are bounded to page size, and
       page size is much smaller than 1 << 32. It's not up to this function to assert that bound
       since we expect this to be a hot path. Moreover it's the kind of hot path where it's hard
       to reduce the amount of work done anywhere else so we might as well not assert too much. */

    result.old_bump = view->bump_offset;
    result.old_bump = (unsigned)pas_round_up_to_power_of_2(result.old_bump, alignment);
```
So, I don't know whether this one is a hot path or not and whether its inputs are small enough.

hyjorc1 replied:
> inputs are unsigned type so seems unnecessary but I can add if you still prefer.

I think `unsigned` can still overflow right?

hyjorc1 replied:
```
unsigned allocation_size = 0xFFFFF000;     // 4,294,963,200 (just below 4GB)
unsigned trailing_padding_size = 0x1000;   // 4,096
unsigned trailing_padding_alignment = 0x2000; // 8,192
```
then
```
rounded = (size + alignment - 1) & -alignment
        = (0xFFFFF000 + 0x1FFF) & ~0x1FFF
        = (0x100000FFF) & 0xFFFFFFFFFFFFE000
        = 0x100000000  // Rounded up to 4,294,967,296
```
So,
```
(lldb) p (void*)(((uintptr_t)0xFFFFF000 + (uintptr_t)0x1FFF) & ~(uintptr_t)0x1FFF)
(void *) 0x0000000100000000
(lldb) p static_cast<unsigned>(0x0000000100000000)
(unsigned int) 0
```

dhecht replied:
Yeah, sorry I wasn't very clear. I just meant that we expect to be doing unsigned 32-bit math on this path given the input and output types. It can overflow, but nothing has changed, I've just made the cast explicit rather than implicit.

Achierius replied:
Agree, there should not be a PAS_ASSERT here -- a PAS_TESTING_ASSERT would make sense but like you said this patch didn't add the code so NP

hyjorc1 reviewed (approved):
r=me

dhecht committed:
[libpas] Fix the libpas tests build
https://bugs.webkit.org/show_bug.cgi?id=293960
rdar://152507524

Reviewed by Yijia Huang.

* Source/bmalloc/libpas/libpas.xcodeproj/project.pbxproj:

pas_thread.{c,h} was added in 295497@main so add it to the
xcode project so that the tests can be built on darwin.

* Source/bmalloc/libpas/src/libpas/pas_segregated_shared_view.h:
(pas_segregated_shared_view_compute_minimum_size_for_bump_with_aligned_padding):

This function was added in 293476@main. The compiler complains about the
implicit conversion when building the tests but not when building the product.
Add the cast to silence the warning  (promoted to error), like we do in other places.

OpenSource/Source/bmalloc/libpas/src/libpas/pas_segregated_shared_view.h:137:9: error:
      implicit conversion loses integer precision: 'uintptr_t' (aka 'unsigned long') to 'unsigned int'
      [-Werror,-Wshorten-64-to-32]
  136 |     return pas_round_up_to_power_of_2(allocation_size, trailing_padding_alignment)
      |     ~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  137 |         + trailing_padding_size;
      |         ^~~~~~~~~~~~~~~~~~~~~~~

Canonical link: https://commits.webkit.org/295762@main
